python/ray/tests/conftest.py

```diff
@@ -397,3 +402,70 @@ def unstable_spilling_config(request, tmp_path):
     ])
 def slow_spilling_config(request, tmp_path):
     yield create_object_spilling_config(request, tmp_path)
+
+
+@pytest.fixture
+def ray_start_chaos_cluster(request):
+    """Returns the cluster and chaos thread.
+
+    Run chaos_thread.start() to start the chaos testing.
+    NOTE: `cluster` is not thread-safe. `cluster`
+    shouldn't be modified by other thread once
+    chaos_thread.start() is called.
+    """
+    os.environ["RAY_num_heartbeats_timeout"] = "5"
+    os.environ["RAY_raylet_heartbeat_period_milliseconds"] = "100"
+    param = getattr(request, "param", {})
+    kill_interval = param.get("kill_interval", 2)
+    # Config of workers that are re-started.
+    head_resources = param["head_resources"]
+    worker_node_types = param["worker_node_types"]
+    timeout = param["timeout"]
+
+    # Use the shutdown RPC instead of signals because we can't
+    # raise a signal in a non-main thread.
+    def kill_raylet(ip, port, graceful=True):
+        raylet_address = f"{ip}:{port}"
+        channel = grpc.insecure_channel(raylet_address)
+        stub = node_manager_pb2_grpc.NodeManagerServiceStub(channel)
+        print(f"Sending a shutdown request to {ip}:{port}")
+        stub.ShutdownRaylet(
+            node_manager_pb2.ShutdownRayletRequest(graceful=graceful))
+
+    cluster = AutoscalingCluster(head_resources, worker_node_types)
+    cluster.start()
+    ray.init("auto")
+    nodes = ray.nodes()
+    assert len(nodes) == 1
+    head_node_port = nodes[0]["NodeManagerPort"]
+    killed_port = set()
+
+    def run_chaos_cluster():
```

Comment:
Hmm what about the design that we discussed earlier about using remote tasks to kill the raylet? Like:
```
@ray.remote
def kill_raylet():
  $ kill -9 local raylet
```

Sorry if I'm forgetting some context that we discussed already, but would this work? Then we wouldn't need to use the RPC because the raylets will get killed by a separate process.

Reply:
For the cluster util, I think it is not easy to find the proper pid that’s not head node. Also, what’s the problem of RPCs here? I feel like this is a cleaner approach than task killing raylet since we can centralize the logic in a single process

Reply:
Hmm I think it will work as long as the head node has 0 CPUs. Then we just submit a kill task that requires 1 CPU.

I'm thinking it would be nice to do it this way because then it doesn't depend on Ray internals. In the physical cluster setting, we can also use the node IP resource to kill a targeted node.

Reply:
For the cluster util setup, it is a little tricky to find the pid of raylet that's not the head node, so I prefer to use this RPC. I think you have a fair point about Ray internals, but I feel like it is unlikely this RPC will have backward incompatible change + it might be more robust to use RPC API (especially when we already have the API) for additional chaos tests in the future? (like RPC delay and stuff like that). If you feel strong about this, I don't have a strong opinion here, so for the nightly test chaos testing, I can just use task that kills the raylet if you feel strongly about it!

Reply:
Okay, sounds good! I think probably either is actually fine for the nightly test, too, whichever you think is simpler.